pkg/odo/cli/component/push.go

```diff
@@ -82,15 +85,60 @@ func (po *PushOptions) Complete(name string, cmd *cobra.Command, args []string)
 
 	// if experimental mode is enabled and devfile is present
 	if experimental.IsExperimentalModeEnabled() && util.CheckPathExists(po.DevfilePath) {
-		envInfo, err := envinfo.NewEnvSpecificInfo(po.componentContext)
+
+		po.Devfile, err = parser.ParseAndValidate(po.DevfilePath)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse devfile")
+		}
+
+		// We retrieve the configuration information. If this does not exist, then BLANK is returned (important!).
+		envFileInfo, err := envinfo.NewEnvSpecificInfo(po.componentContext)
 		if err != nil {
 			return errors.Wrap(err, "unable to retrieve configuration information")
 		}
-		po.EnvSpecificInfo = envInfo
+
+		// If the file does not exist, we should populate the environment file with the correct env.yaml information
+		// such as name and namespace.
+		if !envFileInfo.EnvInfoFileExists() {
+			klog.V(4).Info("Environment file does not exist, creating the env.yaml file in order to use 'odo push'")
+
+			// Since the environment file does not exist, we will retrieve a correct namespace from
+			// either cmd commands or the current default kubernetes namespace
+			namespace, err := retrieveCmdNamespaceDefaultIfNotAvailable(cmd)
+			if err != nil {
+				return errors.Wrap(err, "unable to determine target namespace for devfile")
+			}
+
+			// Retrieve a default name
+			// 1. Use args[0] if the user has supplied a name to be used
+			// 2. If the user did not provide a name, use gatherName to retrieve a name from the devfile.Metadata
+			// 3. Use the folder name that we are pushing from as a default name if none of the above exist
+
+			var name string
+			if len(args) == 1 {
+				name = args[0]
+			} else {
+				name, err = gatherName(po.Devfile, po.DevfilePath)
+				if err != nil {
+					return errors.Wrap(err, "unable to gather a name to apply to the env.yaml file")
+				}
+			}
+
+			// Create the environment file. This will actually *create* the env.yaml file in your context directory.
+			err = envFileInfo.SetConfiguration("create", envinfo.ComponentSettings{Name: name, Namespace: namespace})
```

Comment:
Use the SetComponentSettings here, the create param was removed as part of a recent change